### PR TITLE
Feature: historical TVL

### PR DESCRIPTION
**Pull request by ddallaire**
* Fetch and save TVL values per currency daily
* Keep underlying value and usd value by using USDC as a proxy
* Only save values when the sum is not 0

I added a bunch of guard clauses to shield from null currency values.

### CODE REVIEW

File: src/timeseriesUpdate.ts [hunk 1]
```diff
@@ -73,3 +89,60 @@ export function updateNTokenPresentValueHistoricalData(notional: Notional, curre
   log.debug('Updated nTokenPresentValueHistoricalData variables for entity {}', [nTokenPresentValueHistoricalData.id]);
   nTokenPresentValueHistoricalData.save();
 }
+
+export function updateTvlHistoricalData(notional: Notional, maxCurrencyId: i32, timestamp: i32): void {
+  let tvlCurrencies = new Array<string>();
+  let usdTotal = BigInt.fromI32(0);
+
+  for (let currencyId: i32 = 1; currencyId <= 4; currencyId++) {
```
Comment (jeffywu) [inline src/timeseriesUpdate.ts:97]: This should iterate over max currencies

Comment (ddallaire) [reply src/timeseriesUpdate.ts:97]: Good catch, I added that for testing purposes.

File: schema.graphql [hunk 2]
```diff
@@ -549,6 +549,20 @@ type NTokenPresentValueHistoricalData @entity {
...
+  currency: Currency!
+  underlyingValue: BigInt!
+  usdValue: BigInt!
+}
+
+type TvlHistoricalData @entity {
+  id: ID!
+  timestamp: Int!
+  usdTotal: BigInt!
+  tvlCurrencies: [CurrencyTvl!]!
```
Comment (jeffywu) [inline schema.graphql:563]: Can we rename this from `tvlCurrencies` to `perCurrencyTVL`

File: src/timeseriesUpdate.ts [hunk 3]
```diff
@@ -73,3 +89,60 @@ export function updateNTokenPresentValueHistoricalData(notional: Notional, curre
...
+  for (let currencyId: i32 = 1; currencyId <= maxCurrencyId; currencyId++) {
+    let result = notional.try_getCurrencyAndRates(currencyId);
+    if (!result.reverted) {
+      let assetToken = result.value.value0;
+      let erc20 = ERC20.bind(assetToken.tokenAddress);
+      let assetTokenBalance = erc20.balanceOf(notional._address);
+      let underlyingValue = convertAssetToUnderlying(notional, currencyId, assetTokenBalance);
+      let ethRate = result.value.value2;
+      let ethRateDecimals = result.value.value2.rateDecimals;
+      let ethValue = underlyingValue.times(ethRate.rate).div(ethRateDecimals).div(assetToken.decimals);
```
Comment (jeffywu) [inline src/timeseriesUpdate.ts:106]: I don't believe the second `div` is required. It should just be:
`let ethValue = underlyingValue.times(ethRate.rate).div(ethRateDecimals)`

Comment (ddallaire) [reply src/timeseriesUpdate.ts:106]: If I remove the second division, do I need to add it to the eth to usd conversion?

Comment (jeffywu) [reply src/timeseriesUpdate.ts:106]: I don't believe it is necessary in either, does it cause any issues if its removed?

Comment (ddallaire) [reply src/timeseriesUpdate.ts:106]: No issues, I was having a hard time validating values with kovan. I'll fix this.

**Review by jeffywu — APPROVED** [review]

**Pull request merged**